src/dotnet-serve/DefaultExtensionsMiddleware.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace McMaster.DotNet.Server
+{
+    public class DefaultExtensionsMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly IFileProvider _fileProvider;
+        private readonly DefaultExtensionsOptions _options;
+        private readonly ILogger _logger;
+
+        public DefaultExtensionsMiddleware(RequestDelegate next, IHostingEnvironment hostingEnv, IOptions<DefaultExtensionsOptions> options, ILoggerFactory loggerFactory)
+        {
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            if (hostingEnv == null)
+            {
+                throw new ArgumentNullException(nameof(hostingEnv));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _next = next;
+            _fileProvider = hostingEnv.WebRootFileProvider;
+            _options = options.Value;
+            _logger = loggerFactory.CreateLogger<DefaultExtensionsMiddleware>();
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            if (IsGetOrHeadMethod(context.Request.Method)
+                && !PathEndsInSlash(context.Request.Path))
+            {
+                // Check if there's a file with a matched extension, and rewrite the request if found
+                foreach (string extension in _options.Extensions)
+                {
+                    string filePath = Uri.UnescapeDataString(context.Request.Path.ToString()) + extension;
```

Comment:
Why are you calling Uri.UnescapeDataString? Doesn't the server already normalize context.Request.Path?

Reply:
At one point maybe it didn’t? (copied this from similar middleware I wrote a while ago) - I’ll double-check and remove the call if it’s no longer needed